implement the new handshake

marten-seemann opened the pull request:
Please read https://github.com/libp2p/specs/pull/151.

marten-seemann committed:
implement the new handshake

Stebalien commented on crypto.go:
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling public key failed: %s", err)
+	}
+	certKeyPub, err := x509.MarshalPKIXPublicKey(cert.PublicKey)
```
Is this deterministic?

marten-seemann replied:
Yes, it is: https://golang.org/src/crypto/x509/pkcs1.go?s=3633:3686#L138

Stebalien commented on crypto.go:
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
-func keyToCertificate(sk ic.PrivKey) (crypto.PrivateKey, *x509.Certificate, error) {
-	sn, err := rand.Int(rand.Reader, big.NewInt(1<<62))
+func keyToCertificate(sk ic.PrivKey) (*tls.Certificate, error) {
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
```
Are there any other curves we can use...?

marten-seemann replied:
The Go TLS implementation doesn't support P224, and P256 is curve that produces the smallest signatures. We could also use P384 and P512.

Kubuxu replied:
Does it support `curve25519`?

marten-seemann replied:
`keyToCertificate` does (as well as secp256k1), but the Go standard library doesn't. This was one of the reasons we had to invent this new handshake protocol.

Kubuxu replied:
Ahh, I made a mistake and thought it was `ecdh` not `ecdsa`. (`curve25519` is DH, `ed25519` is signature).

Stebalien commented on crypto.go:
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
+	if err != nil {
+		return nil, err
+	}
+	signature, err := sk.Sign(certKeyPub)
```
Could we add a prefix? Something like `libp2p-tls-handshake:`? For pubsub, we prefix `libp2p-pubsub:`. I'm probably being paranoid but I'd like to make sure we can't possibly sign a cert in this context and have some other system interpret the signature in a different context.

marten-seemann replied:
Sure. This is needs to be defined in the spec though.

Stebalien reviewed (commented):
Nice!

But can we have _lots_ of comments so nobody does something stupid and changes something they shouldn't?

marten-seemann committed:
use ChaCha if one of the peers doesn't have AES hardware support

raulk commented on crypto.go:
```diff
@@ -95,70 +102,95 @@ func getRemotePubKey(chain []*x509.Certificate) (ic.PubKey, error) {
+	}
+	tmpl := &x509.Certificate{
+		SerialNumber: sn,
+		NotBefore:    time.Time{},
```
Should we add some slack to this timestamp to account for clock skew?

marten-seemann replied:
`time.Time{}` is the smallest possible timestamp.

raulk replied:
Ah, I read this as `time.Now()` for some reason. Ok!

raulk commented on crypto.go:
```diff
@@ -1,21 +1,31 @@
-const certValidityPeriod = 180 * 24 * time.Hour
+const certValidityPeriod = 100 * 365 * 24 * time.Hour // ~100 years
+
+var extensionID = getPrefixedExtensionID([]int{1, 1})
```
Should we export the extension ID?

marten-seemann replied:
I don't see why we would need to do that at the moment, so I'd prefer to keep the API as small as possible.
By the way, what do you think about https://github.com/libp2p/specs/pull/151/files#diff-29603a9a6fd077382b3e7ccb9a175726R39?

raulk replied:
See response in https://github.com/libp2p/specs/pull/151/files#r264249948.

raulk commented on transport_test.go:
```diff
@@ -37,17 +41,22 @@ var _ = Describe("Transport", func() {
+			fmt.Fprintf(GinkgoWriter, " using an Ed25519 key: ")
+			priv, _, err = ci.GenerateEd25519Key(rand.Reader)
+		case 3:
+			fmt.Fprintf(GinkgoWriter, " using an Ed25519 key: ")
```
```suggestion
			fmt.Fprintf(GinkgoWriter, " using an secp256k1 key: ")
```

raulk reviewed (commented):
Can't say I'm a crypto expert, but logically LGTM. Minor comments here, and I defer to @Kubuxu and @Stebalien for final approval. Thanks for the great work, @marten-seemann! 🎉

raulk committed:
fix logging when using secp256k1 key in tests

Co-Authored-By: marten-seemann <martenseemann@gmail.com>

marten-seemann committed:
use the new Protocol Labs PEN for the certificate extension

marten-seemann committed:
Merge pull request #23 from libp2p/chacha-preference

use ChaCha if one of the peers doesn't have AES hardware support

marten-seemann committed:
use a prefix when signing the public key

marten-seemann committed:
Merge pull request #26 from libp2p/cert-prefix

use a prefix when signing the public key

marten-seemann committed:
disable session tickets

When resuming a session using session tickets, no certificate chain is
presented, and the callbacks needed to verify the peer identity would
not be called.

marten-seemann commented:
@Kubuxu, @Stebalien I just merged the TLS spec. Can you please review this PR now?

Kubuxu reviewed (commented):
SGWM

Stebalien commented on crypto.go:
```diff
@@ -24,22 +37,21 @@ type Identity struct {
 			VerifyPeerCertificate: func(_ [][]byte, _ [][]*x509.Certificate) error {
 				panic("tls config not specialized for peer")
 			},
+			SessionTicketsDisabled: true,
```
Any reason for this?

marten-seemann replied:
When resuming a session using session tickets, no certificate chain is presented, and the callbacks needed to verify the peer identity would not be called.

We can add session ticket support later (and save the peer ID alongside the session ticket in the session ticket cache).
